src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSearchValues.cs

```diff
@@ -0,0 +1,304 @@
+            description: CreateLocalizableResourceString(nameof(UseSearchValuesDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        protected abstract bool IsConstantByteOrCharArrayVariableDeclaratorSyntax(SyntaxNode syntax, out int length);
+
+        protected abstract bool IsConstantByteOrCharReadOnlySpanPropertyDeclarationSyntax(SyntaxNode syntax, out int length);
+
+        protected abstract bool ArrayFieldUsesAreLikelyReadOnly(SyntaxNode syntax);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffersSearchValues, out _) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlySpan1, out var readOnlySpanType))
+                {
+                    return;
+                }
+
+                var indexOfAnyMethods = GetIndexOfAnyMethods(context.Compilation);
+
+                context.RegisterOperationAction(context => AnalyzeInvocation(context, indexOfAnyMethods, readOnlySpanType), OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(OperationAnalysisContext context, HashSet<IMethodSymbol> indexOfAnyMethodsToDetect, INamedTypeSymbol readOnlySpanType)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+
+            if (!indexOfAnyMethodsToDetect.Contains(invocation.TargetMethod))
+            {
+                return;
+            }
+
+            Debug.Assert(invocation.Arguments.Length is 1 or 2);
+            IArgumentOperation valuesArgument = invocation.Arguments[^1];
+
+            bool isStringIndexOfAny = invocation.TargetMethod.ContainingType.SpecialType == SpecialType.System_String;
+
+            if (isStringIndexOfAny && invocation.Parent is IConditionalAccessOperation)
+            {
+                // We don't flag uses like "string?.IndexOfAny(char[])"
+                // as it's not trivial to rewrite to the span-based variant.
+                return;
+            }
+
+            if (isStringIndexOfAny
+                ? AreConstantValuesWorthReplacingForStringIndexOfAny(valuesArgument.Value)
+                : AreConstantValuesWorthReplacing(valuesArgument.Value, readOnlySpanType))
+            {
+                context.ReportDiagnostic(valuesArgument.CreateDiagnostic(Rule));
+            }
+        }
+
+        private static HashSet<IMethodSymbol> GetIndexOfAnyMethods(Compilation compilation)
+        {
+            var methods = new HashSet<IMethodSymbol>();
```

Comment:
I'm not sure if HashSet is good here. An ImmutableArray might perform better if the collection is expected to be small.

Or maybe this method could be turned to "IsCandidate"?

Reply:
Ok if it's 26 I think it might be reasonable to have a HashSet.